lib/cache.js

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2021 -  Universidad Politécnica de Madrid.
+ *
+ * This file is part of PEP-Proxy
+ *
+ */
+
+const isHex = require('is-hex');
+const debug = require('debug')('pep-proxy:cache');
+const config_service = require('./config_service');
+const NodeCache = require('node-cache');
+const cache = new NodeCache({
+  stdTTL: config_service.get_config().cache_time,
+  checkperiod: config_service.get_config().cache_time
+});
+
+/**
+ * Add a user into the cache
+ */
+function storeUser(token, user) {
+  cache.set(token, { date: new Date(), info: user });
+}
+
+/**
+ *  Add an action+resource+token combo into the cache
+ */
+function storeAction(token, action, resource) {
+  const user = cache.get(token);
+  if (user) {
+    if (!user[action]) {
+      user[action] = [];
+      user[action].push(resource);
+    } else if (user[action] && user[action].indexOf(resource) === -1) {
+      user[action].push(resource);
+    }
+
+    cache.set(token, user);
+  }
+}
+
+/**
+ *  Check if a user is found in the cache
+ *  @return a user if a token for the given user is found and the token
+ *  has not expired
+ */
+function checkTokenCache(token, jwtExpiration, action, resource) {
+  const user = cache.get(token);
+  const config = config_service.get_config();
+  if (!user) {
+    return undefined;
+  }
+  debug('Token found, checking timestamp...');
+  debug(token);
+  const currentTime = new Date().getTime();
+  const tokenTime = token.length <= 40 && isHex(token) ? jwtExpiration * 1000 : user.date.getTime();
```

Comment:
I was trying out your changes locally in my fork and I think I found an issue, the `jwtExpiration * 1000 : user.date.getTime();` must be changed in the conditions, otherwise the cache will not work for Bearer Tokens.

Reply:
Is that a new issue created in the refactoring, or is it an **existing** issue already present in the PEP Proxy? Compare the following:

#### Original [`master`](https://github.com/ging/fiware-pep-proxy/blob/master/lib/idm.js#L158-L194)

```javascript
  const currentTime = new Date().getTime();
  const tokenTime =  token.length <= 40 && isHex(token) ? jwtExpiration * 1000 : cache[token].date.getTime();

   if (currentTime - tokenTime < config.cache_time * 1000) {
```

#### Refactored

```javascript
  const user = cache.get(token);
  const currentTime = new Date().getTime();
  const tokenTime = token.length <= 40 && isHex(token) ? jwtExpiration * 1000 : user.date.getTime();

  if (currentTime - tokenTime > config.cache_time * 1000) {
```

I don't think I've introduced a new bug here, I just changed the code for retrieving a `user` from a memory array  to a cache call.

Reply:
It could be that `token.length <= 40 && isHex(token)` is trying to differentiate bearer tokens from JWT, and that line could be buggy.

Reply:
Ah I see, then it's probably a bug in the existing code.

Reply:
I've added  an additional bearer cache test  76618b0 and run it in debug mode `DEBUG=pep-proxy:* npm t` :

```console
nyc --reporter=html mocha -- --recursive 'test/**/*.js' --reporter spec --timeout 3000 --ui bdd --exit --color true -g 'When the same restricted path is requested multiple times with a bearer token'
```

Returns:

```text
  Authentication: Keyrock IDM
    When the same restricted path is requested multiple times with a bearer token
  pep-proxy:app Starting PEP proxy on port 1026. User authentication via IDM +0ms
  pep-proxy:app Public paths /public +1ms
  pep-proxy:IDM-Client Authenticating user +0ms
  pep-proxy:access Permitted. +0ms
  pep-proxy:IDM-Client Authenticating user +36ms
  pep-proxy:cache Token found, checking timestamp... +0ms
  pep-proxy:cache 111111111 +0ms
  pep-proxy:access Permitted. +19ms
      ✓ should access the user from cache (65ms)
```

So it looks like the cache is being hit.

Reply:
For a bearer token the first two condtions are going to true and then it uses jwtExpiration (which is not defined and therefore 0) to set the timeout value.

Reply:
So should `jwtExpiration !== undefined` be added to the if statement? Something like:

```javascript
if (jwtExpiration) {
 ...  ????
} else 
   tokenTime =  token.length <= 40 && isHex(token) ? jwtExpiration * 1000 : cache[token].date.getTime();
}
```

Reply:
I think that would make sense

Reply:
The more I look at this, the more convinced I am that this is an error in logic.

- `jwtExpiration` is only set if we're using JWT tokens
- A JWT token is long
- A JWT token isn't valid hex - it has full stops in the string

I think the line should be:

```diff
- const tokenTime = token.length <= 40 && isHex(token) ? jwtExpiration * 1000 : user.date.getTime();
+ const tokenTime = jwtExpiration ? jwtExpiration * 1000 : user.date.getTime();
```

i.e. if the JWT token has an expiry use it in preference to the cached expiry value.

Making this change causes one test to fail -  where the cached value says expire but the JWT disagrees the PEP Proxy is currently making an (unnecessary ??) call to the IDM.

Changing the broken test expectation from 2 calls to one fixes the test.

```diff
 idmMock = nock('http://keyrock.com:3000')
        .get('/user?access_token=' + jwt + '&app_id=application_id&action=GET&resource=/restricted')
-      .times(2)
+      .times(1)
        .reply(StatusCodes.OK, keyrock_permit_response);
```

@aalonsog  @apozohue10 - should this logic be updated?

Reply:
Sorry for the delay in answering. This is definitely an error in logic and what you propose makes totally sense. I will test this PR and then I will update the code to include the logic change. Thank you very much